modulus/models/module.py

```diff
@@ -65,55 +106,286 @@ def debug(self):
         # TODO: set up debug log
         # fh = logging.FileHandler(f'modulus-core-{self.meta.name}.log')
 
-    def save(self, file_name: Union[str, None] = None) -> None:
+    def save(
+        self, file_name: Union[str, None] = None, save_git_hash: bool = False
+    ) -> None:
         """Simple utility for saving just the model
 
         Parameters
         ----------
         file_name : Union[str,None], optional
             File name to save model weight to. When none is provide it will default to
             the model's name set in the meta data, by default None
+        save_git_hash : bool, optional
+            Whether to save the git hash of the current commit, by default False
 
         Raises
         ------
         IOError
             If file_name provided has a parent path that does not exist
         """
+
+        if file_name is not None and not file_name.endswith(self._file_extension):
+            raise ValueError(
+                f"File name must end with {self._file_extension} extension"
+            )
+
+        directory = Path(file_name)
+        if not directory.parents[0].is_dir():
+            raise IOError(
+                f"Model checkpoint parent directory {directory.parents[0]} not found"
+            )
+        if not directory.is_dir():
+            os.makedirs(directory)
+
+        torch.save(self.state_dict(), directory.joinpath("model.pt"))
+
+        with open(directory.joinpath("args.json"), "w") as f:
+            json.dump(self._args, f)
+            if file_name is None:
+                file_name = self.meta.name + ".pt"
+
+        # Save the modulus version and git hash (if available)
+        metadata_info = {"modulus_version": modulus.__version__,
+                         "mdlus_file_version": self.__version__}
+
+        if save_git_hash:
+            import git
+            repo = git.Repo(search_parent_directories=True)
+            try:
+                metadata_info["git_hash"] = repo.head.object.hexsha
+            except git.InvalidGitRepositoryError:
+                metadata_info["git_hash"] = None
+
+        with open(directory.joinpath("metadata.json"), "w") as f:
+            json.dump(metadata_info, f)
```

Comment:
Can we use fsspec abstraction here instead of just posix file writes? Something like:
```python
from modulus.utils.filesystem import _get_fs

path_str = str(directory.joinpath("metadata.json"))
fs = _get_fs(path_str)
fs.pipe(path_str, json.dumps(metadata_info))
```
https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem.pipe

Reply:
The other option would be to write everything to a temp directory and then copy to the final destination:
```python
with tempfile.TemporaryDirectory() as tmpdirname:
    with open(os.join(tmpdirname, "metadata.json"), "w") as f:
            json.dump(metadata_info, f)
    ...
    fs.put(tmpdirname, directory, recursive=True)
```